torch/_inductor/memory.py

```diff
@@ -0,0 +1,555 @@
+                    sched_buf.mpi_buffer.size_alloc += buf.mpi_buffer.size_alloc
+                    buf.mpi_buffer.size_alloc = 0
+
+
+def map_successor_nodes_with_predecessor_buffers(
+    nodes: List[BaseSchedulerNode],
+    name_to_input_buf: Dict[str, FreeableInputBuffer],
+    name_to_buf: Dict[str, SchedulerBuffer],
+) -> None:
+    """
+    For scheduling and memory estimation, for each scheduler node, we maintain
+    a list of its dependency buffers (SchedulerBuffer and FreeableInputBuffer).
+    This is similar to node.read_writes.reads, which is a list of Dep.
+    Reversely, for each SchedulerBuffer / FreeableInputBuffer, assign its successor nodes.
+    A buffer's successor nodes determines when a buffer can be freed.
+    """
+    for node in nodes:
+        node.mpi_node = MemoryPlanningInfoForNode()
+        node.mpi_node.pred_buffers = []
+        for dep_name in {dep.name for dep in node.unmet_dependencies}:
+            sched_buf = name_to_buf.get(dep_name)
+            if sched_buf:
+                node.mpi_node.pred_buffers.append(sched_buf)
+                sched_buf.mpi_buffer.succ_nodes.add(node)
+        for dep_name in {
+            dep.name for dep in node.read_writes.reads - node.unmet_dependencies
+        }:
+            input_buf = name_to_input_buf.get(dep_name)
+            if input_buf:
+                node.mpi_node.pred_buffers.append(input_buf)
+                input_buf.mpi_buffer.succ_nodes.add(node)
+
+
+def estimate_peak_memory(
+    nodes: List[BaseSchedulerNode],
+    name_to_input_buf: Dict[str, FreeableInputBuffer],
+    graph_outputs: Set[str],
+) -> Tuple[int, List[int]]:
+    """
+    Given a list of nodes in their execution order, estimate the peak memory, by
+    keeping track of the liveliness of SchedulerBuffers and FreeableInputBuffers.
+
+    Returns:
+        int: peak memory
+        List[int]: memory usage at each node.
+    """
+
+    # map each scheduler buffer to its size, start time, and end time
+    @dataclasses.dataclass
+    class BufferInfo:
+        buffer: Union[SchedulerBuffer, FreeableInputBuffer]
+        size_alloc: int
+        size_free: int
+        start_time: int
+        end_time: int
+
+    name_to_buf_info: Dict[str, BufferInfo] = {}
+    node_name_to_time: Dict[str, int] = {}
+
+    # assign start_time
+    for buf_name, input_buf in name_to_input_buf.items():
+        name_to_buf_info[buf_name] = BufferInfo(
+            input_buf,
+            input_buf.mpi_buffer.size_free,
+            input_buf.mpi_buffer.size_free,
+            0,
+            0,
+        )
+    for t, node in enumerate(nodes):
+        node_name_to_time[node.get_name()] = t
+        for sched_buf in node.get_outputs():
+            name_to_buf_info[sched_buf.get_name()] = BufferInfo(
+                sched_buf,
+                sched_buf.mpi_buffer.size_alloc,
+                sched_buf.mpi_buffer.size_free,
+                t,
+                t,
+            )
```

Comment:
IMO, It makes it a little more difficult to follow to have intermediary states which are not correct. Here, the end_time is not right for a period of time. I would lean more towards having intermediary objects as you need them instead of incrementally updating objects such that there is a period of time where they are not correct. Kind of similar also to the dataclasses comment.

Reply:
I restructured the code here to get rid of these "intermediary states". (these are interesting comments, I need to think a bit about how to address similar comments for dataclasses as quite some restructuring is needed)